lib/dynflow/exporters/abstract.rb

```diff
@@ -0,0 +1,104 @@
+module Dynflow
+  module Exporters
+    class Abstract
+
+      attr_reader :index
+
+      def self.export_execution_plan_id(world, execution_plan_id, options = {})
+        self.new(world, options)
+          .add_id(execution_plan_id)
+          .finalize.index[execution_plan_id][:result]
+      end
+
+      def self.export_execution_plan(execution_plan, options = {})
+        self.new(execution_plan.world, options)
+          .add(execution_plan)
+          .finalize.index[execution_plan.id][:result]
+      end
+
+      # Expected workflow:
+      # Exporter.new.add(foo).add(bar).add(baz).finalize.result
+      def initialize(world = nil, options = {})
+        @world   = world
+        @options = options
+        @index   = {}
+      end
+
+      # Add the execution_plan to the index, thus queueing it for exporting
+      def add(execution_plan)
+        @index[execution_plan.id] = {
+          :plan   => execution_plan,
+          :result => nil
+        }
+        self
+      end
+
+      # Add provided id to the index
+      # The id will be resolved to the execution plan
+      # in the #resolve_ids method along with all the other
+      # index entries whose :plan is nil
+      def add_id(execution_plan_id)
+        @index[execution_plan_id] = {
+          :plan   => nil,
+          :result => nil
+        }
+        self
+      end
+
+      # The same as #add, but takes Array[ExecutionPlan]
+      def add_many(execution_plans)
+        execution_plans.each { |plan| add plan }
+        self
+      end
+
+      # The same as #add_id, but takes Array[String]
+      def add_many_ids(execution_plan_ids)
+        execution_plan_ids.each { |plan_id| add_id plan_id }
```

Comment:
This place doesn't interact with the database at all, everything thouching the db happens [here](https://github.com/Dynflow/dynflow/pull/115/files#diff-4ca04e4243569cbfd52809ce04bd1743R97). Dynflow uses Sequel (instead of ActiveRecord) to work with the database and in this case the retrievals happen in batches of 1000 records afaik.

Introducing AR would be quite big change